### ISSUE TITLE
IDatasetForm does not provide package_type to setup_template_variables

### ISSUE THREAD

**Issue by ccancellieri** [ckan/ckan#6518]
**CKAN version**
2.8.x

**Describe the bug**
I'm creating a new plugin to manage a set of types let's say:

```
def package_types(self):
    return ['type1',' type2']
```

so the ckan core function:

```
class PackageController(base.BaseController):
      ...
      def _setup_template_variables(self, context, data_dict, package_type=None):
      return lookup_package_plugin(package_type).\
      setup_template_variables(context, data_dict)
      ...

```

Will lookup for my plugin and based on **package_type** assign the management of that type to my plugin....

But then the package_type is not passed to my plugin... so.

How I can get the type of the package we are creating not looking at the url ?

 http://{CKAN_BASE}/{TYPE}/new

Context is not shipping this info and the IDatasetForm does not ship it as param of setup_template_variables... ?

**Expected behavior**

I would expect to have the type which is selecting my plugin.

So I can differentiate the view to provide.

If we suppose to have 1 to 1 match to IDatasetForm and the provided format,

which is the point of return multiple types with *package_types()* if we can't know which type we are going to render?

I think we may:
- pass the type into the context
- change the prototype adding the package_type to the method sign

**Comment by smotornyuk** [comment]
I think that plugin that defines multiple package types has to provide some generic interface for **every** package type it defines. 
If you expect different behavior for different package types - just register every separate type inside different plugin

**Comment by ccancellieri** [comment]
> I think that plugin that defines multiple package types has to provide some generic interface for **every** package type it defines. If you expect different behavior for different package types - just register every separate type inside different plugin

So the match is 1 plugin 1 type... which come back to the question... so what's the point of returning a set of types?

**Comment by smotornyuk** [comment]
You can register multiple package types in one go, that's the point. As long as all of them share the same logic of `setup_template_variables`, you can do it. 
Let's imagine you developing the `report` extension, that provides some sort of `wheater`, `population`, `economics` reports(as a separate package type). All of these package-report types share the same template variables(maybe start/end date). So you can define them via a single plugin.

And whenever you need different logic for different types, you go with the 1-to-1 scheme. It has sense because otherwise you will define a lot of non-related package types inside a single plugin and end up with huge unmaintainable plugin


> why plugin should not be informed 

There is no real reason. I definitely wouldn't mind against it if it was a new interface. I love freedom.
But it's an old interface and initially, it was implemented in this way. As you said, now it requires changes and will break the existing code. So we come to the question **do we really need this change? Does it worth breaking existing extensions?**

I think - no. As I mentioned, mixing package types with different logic will only make plugins more complex. Breaking change that allows writing bad code is not what I want:)

**Comment by smotornyuk** [comment]
PS I am not really against this change, just to make it clear, and I will support PR that changes this behavior. But personally, I think that this will be "the change that only allows to write worse code@

**Comment by ccancellieri** [comment]
Yes, I see your point but I don't understand why it should be hidden.

I can't see the point of hide an information which is available to the backend to the template.

To give a bit of context.

I'm writing a jsonschema plugin which is by it'self extensible. The extensions are provided in form of json and jsonschema and no code should be provided.

I also would like to allow people to render their own type not reimplementing the whole plugin but only providing a template based on the type.

The list of types is defined by the administrator providing json schema and json templates and (optionally a jinja2 template).

As you see my plugin allow to render all the provided types but I don't see why the type of the rendered plugin should be lost.

**Comment by ccancellieri** [comment]
by the way I'm also providing several options making this improvement possible NOT breaking any interface.

**Comment by smotornyuk** [comment]
You didn't get it:) 
It **should not** be hidden, it **is** hidden. I.e, there are no reasons to hide this information, but that's how it's done at the moment. That's all. I am not arguing for hiding it, I rather explain why I have to intention to fix it myself :laughing:  As long as you or someone else will work on PR - you're welcome.

In case you only need different templates for different package types, [there is a way](https://github.com/ckan/ckan/blob/2.9/ckan/views/dataset.py#L57-L65) to do it in CKAN v2.9. You don't have to modify `setup_template_variables`. Instead, just define `IDatasetForm.new_template/read_template/etc` not as `def new_template(self):...`, but as `def new_template(self, package_type):...` and you'll get the type

>  I'm also providing several options 

Yep, that's a good point. I don't like the idea of using data_dict, because it may be confused with the data sent by user. But `context` is quite a good candidate for this case.

**Comment by amercader** [comment]
So, to sum up the discussion on the dev meeting: `setup_template_variables()` is unreliable as it does not get consistently calls on all endpoints (eg when creating a new resource), but it makes sense that plugins get the `package_type` passed to them.
At the same time we need to maintain compatibility with older plugins that don't expect this parameter so here's what needs doing:

1. Add a new `package_type` parameter to the `setup_template_variables()` hook: https://github.com/ckan/ckan/blob/bcf2941ea9b46e064ae7a9e12d82662f10a176d2/ckan/plugins/interfaces.py#L1185
2. Pass it in the place where this hook is called in the dataset blueprint ([here](https://github.com/ckan/ckan/blob/bcf2941ea9b46e064ae7a9e12d82662f10a176d2/ckan/views/dataset.py#L49)), but only if the plugin accepts this new param. You can use the same logic used in the `_get_pkg_template()`  method to see if the hook accepts the parameter or not:
https://github.com/ckan/ckan/blob/bcf2941ea9b46e064ae7a9e12d82662f10a176d2/ckan/views/dataset.py#L54-L61
